src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -207,13 +208,43 @@ public Boolean getUserProperty(String userId, String recipeId, String mapName, T
   public List<RecipeMetadata> getRecipesMatchingTags(
       List<String> tagIds, SortingMethod sortingMethod, int page) {
     Query recipesQuery = recipesMatchingTags(tagIds, tagIds.iterator());
-    return getRecipeMetadataQuery(recipesQuery, sortingMethod, page);
+    List<RecipeMetadata> results = getRecipeMetadataQuery(recipesQuery, SortingMethod.NONE, page);
+
+    switch (sortingMethod) {
+      case TOP:
+        System.out.println("Sorting by: TOP");
+        Collections.sort(
+            results, Collections.reverseOrder(Comparator.comparingLong(RecipeMetadata::getVotes)));
```

Comment:
Why is this necessary when `getRecipeMetadataQuery` already sorts and uses the firestore api to sort?

Reply:
This is where the Firestore concept of a custom index comes into play. If we make a request filtering by the contents of the _tagIds_ field, while sorting on the _votes_ or _timestamp_ field, the request fails and Firestore prompts you to create a custom index. However, a custom index only works to resolve this problem for a single tag and sorting field combo (ie _tagIds.DiabfQ87RYNP6Ehre8Oi_ and _timestamp_). This is insufficient for the flexibility we want for a user searching by any combination of tags